NOM-7 SSO documentation (#277)
neo4j/docs-ops-manager: modules/ROOT/pages/installation/sso.adoc

= Single Sign-On (SSO) Configuration
:description: This section describes the configuration steps to enable Single Sign-On (SSO) for NOM server

== Overview

NOM supports SSO for authentication using external Identity Providers (IdP) that support OAuth2/OIDC implicit authentication flow.
Currently, Microsoft Entra ID (formerly Azure Active Directory), Google Identity and Okta are supported.
More IdP types and alternative authentication flows may be added in future NOM releases.
SSO can be used as the only authentication method or in combination with username/password login.
When both methods are enabled, users can choose the login method on the NOM login page.

== Configuration

One or more SSO configs can be set up containing the options needed to perform SSO authentication with an external Identity Provider (IdP).

Configurations are identified by a name, which is used to reference the configuration in the environment variable names or command line arguments.
The name can contain only letters and digits. In the reference below, the placeholder `<CONFIG>` is used to indicate where the configuration name should be used.

> The configuration contains options needed to obtain an access token on the NOM frontend side, validate the access token on the NOM backend side and map token claims to an SSO user in NOM, including assign a NOM role to the SSO user.

Reviewer: ```suggestion
The configuration contains options needed to obtain an access token on the NOM frontend side, validate the access token on the NOM backend side and map token claims to an SSO user in NOM, including assigning a NOM role to the SSO user.
```

Reviewer: ```suggestion
The configuration contains options needed to obtain, validate and map access tokens to SSO users in NOM, including assigning NOM roles to SSO users.
```

Reviewer: Maybe the current sentence is a bit too technical overall?

> Only the options needed for obtaining the token are send to the frontend.

Reviewer: ```suggestion
Only the options needed for obtaining the token are sent to the frontend.
```

Configurations are validated during the startup of NOM server.
If a configuration is invalid, the errors are reported in the logs.

=== Common options

[cols="<,<,<, <",options="header"]
|===
| Command line argument
| Environment variable name
| Description
| Example value

| `app.sso.active-configurations`
| `APP_SSO_ACTIVE_CONFIGURATIONS`
| Lists comma-separated names of configurations that should be offered as login options on the NOM login page.
| `myConfig1,myconfig2`

| `app.username-password-login.enabled`
| `APP_USERNAME_PASSWORD_LOGIN_ENABLED`
| Username/password login can be enabled or disabled, enabling SSO login either be offered additionally or to be the only login method.
If disabled, the pages "NOM Administrators" and "Profile" will not be available in the NOM UI.
| `false`

| `app.sso.configurations.<CONFIG>.idp-type`
| `APP_SSO_CONFIGURATIONS_<CONFIG>_IDP_TYPE`

> | Type of the Identity Provider (IdP). Currently only `ENTRA_ID`, `GOOGLE_IDENTITY` and `OKTA` are supported.

Reviewer: ```suggestion
| Type of the Identity Provider (IdP). Currently `ENTRA_ID`, `GOOGLE_IDENTITY` and `OKTA` are supported.
```

| `['given_name']`

| `app.sso.configurations.<CONFIG>.lastname-claim`
| `APP_SSO_CONFIGURATIONS_<CONFIG>_LASTNAME_CLAIM`
| SpEL expression to override the default field for the last name in the token claims.
| `['family_name']`
|===

[#_role_assignment]
==== Role assignment

The `admin-check` and `read-only-user-check` options are https://docs.spring.io/spring-framework/reference/core/expressions/language-ref.html[SpEL expressions] that are evaluated against the claims in the access token obtained from the IdP.

The claims object is available as the root object in the expression context.

The expressions should evaluate to `true` or `false`.
If the `admin-check` expression evaluates to `true`, the user is assigned the NOM Admin role in NOM.
Otherwise, the read-only-user-check expression is evaluated.
If it evaluates to `true`, the user is assigned the read-only role in NOM.
If both expressions evaluate to `false`, the user is denied access to NOM.

The permissions of the read-only role include:

* Viewing DBMS status
* Viewing DBMS metrics
* Viewing and downloading DBMS query logs
* Viewing agents and their details
* Viewing the Upgrades page

> * Read-only access to the DBMS Security page

Reviewer: Maybe better to remove this list and reference https://neo4j.com/docs/ops-manager/current/nom-user-management/#_roles_of_nom_users

Author: I like referencing stuff to avoid redundancies. And we should add the link for the NOM Admin role description.

But the `Administrator` mentioned there is not exactly the "read-only role" described above, because `Administrator` can potentially have write permissions.

Or were you thinking about explaining it additionally?

Reviewer: Ah right. I didn't realize. But maybe it makes sense to consolidate the role info on that page anyway? If I understand correctly, the read-only role mentioned here is like a non-configurable `Administrator` with a read-only access on *.

To debug issues with role assignment, see the xref:_troubleshooting[Troubleshooting section] below.

==== Overriding claim fields

By default, IdP specific fields in the token claims are used to create or update the SSO user fields in NOM.
It is possible to override the default fields by providing SpEL expressions using the options `username-claim`, `email-claim`, `firstname-claim` and `lastname-claim`.

The expressions should evaluate to a string value and are evaluated against the claims object in the access token.
If an expression evaluates to `null` or if there are errors during evaluation, the corresponding user field is set according to the default behavior.

Debugging issues with claim field extraction can be done as described in the xref:_troubleshooting[Troubleshooting section] below.

> === Microsoft Entra ID

Reviewer 2: May be move individual IdP config docs to their own pages?

Author: let us refactor that later when we add docs for the other IdPs. For the beta it should be ok I guess.

NOM server supports Single Sign-On (SSO) authentication using Microsoft Entra ID (formerly Azure Active Directory) using implicit authentication flow.

> ==== Set up Entra ID Application

Reviewer: We've gotten feedback for log forwarding that we shouldn't document third party systems. But in that case the setup is so complicated that noone would manage a successful config elsewise.

In this case, I would *hope* an admin/IT staff responsible for enrolling SSO applications would know how to set it up on their side? We could put this info in a Google doc in the meantime, to avoid your hard work going to waste in case State farm requests it, or if we do decide we really need to. But that's just my take.

Author: Yes, lets put it into a Google doc that we'll send along with the beta to the customer.

Reviewer 2: Sometimes there are several ways to achieve the same effect from an IdP config and we should also not be too opinionated on what we suggest because we prefer it that way.

I think we should only set expectations on our config & inputs and let the customers figure out the best way to provide those.

Author: Ok, I'll add something about that in the doc. E.g. "the following steps _can_ be used to create a working SSO config for NOM"

Author: see 
15cbc1b. I moved the concrete steps to [this Google doc](https://docs.google.com/document/d/1K1sr_FjAj9UJWAnKghHxjiHx8tF7i0aOZSDxZ54uBmI/edit?tab=t.0).

| `APP_SSO_CONFIGURATIONS_<CONFIG>_AUTHORITY`
| Authority URL of the Entra ID tenant.
It can be found in the "Endpoints" section of the application page.
| `https://login.microsoftonline.com/ddd85725-ed2a-49a4-a19e-11c8d29f9a0f`

| `app.sso.configurations.<CONFIG>.issuer`
| `APP_SSO_CONFIGURATIONS_<CONFIG>_ISSUER`
| Issuer URL for validating the access token.
It is usually derived from "Directory (tenant) ID" of the Entra ID app.
| `https://sts.windows.net/ddd85725-ed2a-49a4-a19e-11c8d29f9a0f/` (v1.0 tokens) or `https://login.microsoftonline.com/ddd85725-ed2a-49a4-a19e-11c8d29f9a0f/v2.0` (v2.0 tokens)

| `app.sso.configurations.<CONFIG>.audience`
| `APP_SSO_CONFIGURATIONS_<CONFIG>_AUDIENCE`
| Audience for validating the access token.
It is usually the same as the "Application ID URI" shown on the "Expose an API" page in Entra ID. (optional)
| `api://ccc0e55f-5f3e-4da8-9002-2830225aae1b` (v1.0 tokens) or `ccc0e55f-5f3e-4da8-9002-2830225aae1b` (v2.0 tokens)
|===

==== Example configuration

The following is an example configuration using environment variables to set up a single SSO configuration named `ENTRA` and disable username/password login.

[IMPORTANT]
====

> Make to quote variable values containing special characters such as `?` to prevent the shell from interpreting them.

Reviewer 2: ```suggestion
Make sure to quote variable values containing special characters such as `?` to prevent the shell from interpreting them.
```

Author: 04d7779

One or more SSO configs can be set up containing the options needed to perform SSO authentication with an external Identity Provider (IdP).

Configurations are identified by a name, which is used to reference the configuration in the environment variable names or command line arguments.

> The name can be freely chosen, but should only contain letters and digits.

Reviewer: Do we think "alphanumeric" is too advanced a term to spring on our users? 😄

Author: 04d7779 - better?
